Mathlib/Topology/Algebra/Module/Basic.lean

```diff
@@ -0,0 +1,2695 @@
+  rw [zero_smul, add_zero] at this
+  obtain ⟨_, hu : y + _ • _ ∈ s, u, rfl⟩ :=
+    nonempty_of_mem (inter_mem (Filter.mem_map.1 (this hy)) self_mem_nhdsWithin)
+  have hy' : y ∈ ↑s := mem_of_mem_nhds hy
+  rwa [s.add_mem_iff_right hy', ← Units.smul_def, s.smul_mem_iff' u] at hu
+#align submodule.eq_top_of_nonempty_interior' Submodule.eq_top_of_nonempty_interior'
+
+variable (R M)
+
+/-- Let `R` be a topological ring such that zero is not an isolated point (e.g., a nontrivially
+normed field, see `NormedField.punctured_nhds_neBot`). Let `M` be a nontrivial module over `R`
+such that `c • x = 0` implies `c = 0 ∨ x = 0`. Then `M` has no isolated points. We formulate this
+using `NeBot (𝓝[≠] x)`.
+
+This lemma is not an instance because Lean would need to find `[ContinuousSMul ?m_1 M]` with
+unknown `?m_1`. We register this as an instance for `R = ℝ` in `Real.punctured_nhds_module_neBot`.
+One can also use `haveI := Module.punctured_nhds_neBot R M` in a proof.
+-/
+theorem Module.punctured_nhds_neBot [Nontrivial M] [NeBot (𝓝[≠] (0 : R))] [NoZeroSMulDivisors R M]
+    (x : M) : NeBot (𝓝[≠] x) := by
+  rcases exists_ne (0 : M) with ⟨y, hy⟩
+  suffices : Tendsto (fun c : R => x + c • y) (𝓝[≠] 0) (𝓝[≠] x); exact this.neBot
+  refine' Tendsto.inf _ (tendsto_principal_principal.2 <| _)
+  · convert tendsto_const_nhds.add ((@tendsto_id R _).smul_const y)
+    rw [zero_smul, add_zero]
+  · intro c hc
+    simpa [hy] using hc
+#align module.punctured_nhds_ne_bot Module.punctured_nhds_neBot
+
+end
+
+section LatticeOps
+
+variable {ι R M₁ M₂ : Type _} [Semiring R] [AddCommMonoid M₁] [AddCommMonoid M₂] [Module R M₁]
+  [Module R M₂] [u : TopologicalSpace R] {t : TopologicalSpace M₂} [ContinuousSMul R M₂]
+  (f : M₁ →ₗ[R] M₂)
+
+theorem continuousSMul_induced : @ContinuousSMul R M₁ _ u (t.induced f) := by
+  let _ : TopologicalSpace M₁ := t.induced f
+  refine' ⟨continuous_induced_rng.2 _⟩
+  simp_rw [Function.comp, f.map_smul]
+  exact continuous_fst.smul (continuous_induced_dom.comp continuous_snd)
+#align has_continuous_smul_induced continuousSMul_induced
+
+end LatticeOps
+
+namespace Submodule
+
+variable {α β : Type _} [TopologicalSpace β]
+
+instance continuousSMul [TopologicalSpace α] [Semiring α] [AddCommMonoid β] [Module α β]
+    [ContinuousSMul α β] (S : Submodule α β) : ContinuousSMul α S :=
+  continuousSMul_induced S.subtype
+#align submodule.has_continuous_smul Submodule.continuousSMul
+
+instance topologicalAddGroup [Ring α] [AddCommGroup β] [Module α β] [TopologicalAddGroup β]
+    (S : Submodule α β) : TopologicalAddGroup S :=
+  inferInstanceAs (TopologicalAddGroup S.toAddSubgroup)
+#align submodule.topological_add_group Submodule.topologicalAddGroup
+
+end Submodule
+
+section closure
+
+variable {R R' : Type u} {M M' : Type v} [Semiring R] [Ring R']
+  [TopologicalSpace M] [AddCommMonoid M] [TopologicalSpace M'] [AddCommGroup M'] [Module R M]
+  [ContinuousConstSMul R M] [Module R' M'] [ContinuousConstSMul R' M']
+
```

Comment:
What's happened is that @urkud's initial fix of the file (which @mcdoll then merged at the beginning) deleted the two `closure_smul_self_*` theorems and added the two new `smul_closure` theorems in their place. I suspect this is because `closure_smul_self_subset` runs into an intractable problem at _its_ start: the instance `TopologicalSpace (R × M)` can't be synthesised. Is replacing theorems to ease Lean's task OK?

Reply:
I've also generalized some lemmas from `[ContinuousSMul]`  to `[ContinuousConstSMul]`. I'll backport these changes to Lean 3 tomorrow. I was going to do it earlier and forgot. Sorry about this.